src/OrleansRuntime/Catalog/Catalog.cs

```diff
@@ -632,46 +632,43 @@ private void CreateGrainInstance(string grainTypeName, ActivationData data, stri
 
             GrainTypeData grainTypeData = GrainTypeManager[grainClassName];
 
+            //Get the grain's type
             Type grainType = grainTypeData.Type;
 
-            //Create a new instance of the given grain type
-            var grain = grainCreator.CreateGrainInstance(grainType, data.Identity);
-            grain.Data = data;
-
+            //Gets the type for the grain's state
             Type stateObjectType = grainTypeData.StateObjectType;
-            object state = null;
-            if (stateObjectType != null)
-            {
-                state = Activator.CreateInstance(stateObjectType);
-            }
             
             lock (data)
             {
-                data.SetGrainInstance(grain);
-                var statefulGrain = grain as IStatefulGrain;
-                if (statefulGrain != null)
+                Grain grain;
+
+                //Create a new instance of a stateless grain
+                if (stateObjectType == null)
                 {
-                    if (state != null)
-                    {
-                        SetupStorageProvider(data);
-                        statefulGrain.GrainState.State = state;
-                        statefulGrain.SetStorage(new GrainStateStorageBridge(data.GrainTypeName, statefulGrain,
-                            data.StorageProvider));
-                    }
+                    //Create a new instance of the given grain type
+                    grain = grainCreator.CreateGrainInstance(grainType, data.Identity);
```

Comment:
I'm a little bit concerned that you moved grain instantiation inside the lock. It might be harmless in some situations, but if using DI, then the instantiation process might be doing extra stuff that we haven't accounted for. What was the reason to change this?